storagenode/pieces/readwrite.go

```diff
@@ -79,16 +193,80 @@ func NewReader(blob storage.BlobReader) (*Reader, error) {
 	if err != nil {
 		return nil, Error.Wrap(err)
 	}
+	formatVersion := blob.GetStorageFormatVersion()
+	if formatVersion >= storage.FormatV1 {
+		if size < V1PieceHeaderReservedArea {
+			return nil, Error.New("invalid piece file for storage format version %d: too small for header (%d < %d)", formatVersion, size, V1PieceHeaderReservedArea)
+		}
+		size -= V1PieceHeaderReservedArea
+	}
 
-	reader := &Reader{}
-	reader.blob = blob
-	reader.size = size
-
+	reader := &Reader{
+		formatVersion: formatVersion,
+		blob:          blob,
+		size:          size,
+	}
 	return reader, nil
 }
 
+// GetStorageFormatVersion returns the storage format version of the piece being read.
+func (r *Reader) GetStorageFormatVersion() storage.FormatVersion {
+	return r.formatVersion
+}
+
+// GetPieceHeader reads, unmarshals, and returns the piece header. It may only be called
+// before any Read() calls. (Retrieving the header at any time could be supported, but for
+// the sake of performance we need to understand why and how often that would happen.)
+func (r *Reader) GetPieceHeader() (*pb.PieceHeader, error) {
+	if r.formatVersion < storage.FormatV1 {
+		return nil, Error.New("Can't get piece header from storage format V0 reader")
+	}
+	if r.pos != 0 {
+		return nil, Error.New("GetPieceHeader called when not at the beginning of the blob stream")
+	}
+	// We need to read the size of the serialized header protobuf before we read the header
+	// itself. The headers aren't a constant size, although V1PieceHeaderReservedArea is
+	// constant. Without this marker, we wouldn't have any way to know how much of the
+	// reserved header area is supposed to make up the serialized header protobuf.
+	var headerBytes [V1PieceHeaderReservedArea]byte
```

Comment:
Rather than allocating the entire reserved area here (which we're assuming will always be more than we need), why not allocated the framing bytes, then read the header size, and _then_ allocate the size we need for the header bytes? Would splitting it into two allocations be slower even though it would always be a smaller allocation?

Reply:
Yes, two allocations would be slower, plus I don't want to have to write `var headerBytes [V1PieceHeaderReservedArea - v1PieceHeaderFramingSize]` and have to think about whether that's correct each time.

Reply:
Actually, since I just saw this again, it probably wouldn't be slower, since this is a constant size and I think Go would just put them all on the stack. But I think that change would still be too much thinking for not enough benefit at this phase :)